### PR TITLE
ARC60 update to emphasize schemas 

### PR DESCRIPTION
**Pull request by k13n**
This PR puts more emphasis on scopes and schemas in ARC60 and it tries to simplify the API.

Some changes:
- There's a new Section "Scopes and Schemas" that explains what a schema is. A schema needs to define three functions, (`parse`, `validate`, and `serialize`) that allow a wallet to parse the structured data, validate it against a set of rules, and serialize it into a byte array that can be signed
- Dropped the `StdSignMetadata` interface and moved all its relevant fields to `StdSignData`. To me it was not clear what belongs to one  or the other, hence merging them made more sense to me.
- Dropped the `encoding` from the `StdSignMetadata`. The schema defines how the data is decoded.

This PR includes all of @ehanoc's work on a reference implementation for ARC60. The reference implementation is updated to reflect the focus on the schemas.

### CODE REVIEW

File: ARCs/arc-0060.md [hunk 1]
```diff
@@ -39,111 +37,143 @@ The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL
...
-  "scope": "...",
-  "schema": "/{.../}",
-  "encoding": "..."
+  },
+  "scope": "MX_RANDOM"
 }
 ```
 
-### Interfaces
+The schema of the structured data (as identified by the `scope`) allows wallets to interpret the data, validate it, present it to the user, and serialize it into a byte string that can be signed. To this end, a schema must define three functions that wallets can use to process the data:
```
Comment (deanstef) [inline ARCs/arc-0060.md:77]: Are we sure it is the responsibility of the schema to define these three functions? IMO ARCs must define them.

File: ARCs/arc-0060.md [hunk 2]
```diff
@@ -39,111 +37,143 @@ The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL
...
+- `parse`: This function turns the input `data` (a string) into a structured object (e.g., a fixed-sized byte array, a JSON object, etc)
+- `validate`: This function validates the input against a known schema for this scope and determines if an input is safe to sign
+- `serialize`: This function deterministically serializes the structured data into a byte array that can be signed
+
+Ultimately, the `signData` function returns the signature `Sig(serialize(parse(data)), sk)` where `sk` is the `signer`'s signing key.
```
Comment (deanstef) [inline ARCs/arc-0060.md:82]: 64-bytes signature?

File: ARCs/arc-0060.md [hunk 3]
```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
...
-The `ScopeType` enumerates constant strings with the scope of a signing action.
+The `ScopeType` enumerates constant strings with the scope of a signing action. Any extension of this ARC **SHOULD** adopt an existing `ScopeType`, or introduce a new one.
+
+This ARC introduces two scope types:
 
-This ARC introduces two scope types.
+| ScopeType          | Description |
+| -----------------  | ----------- |
+| MX_RANDOM          | Random data but it's prefixed with the domain separator `MX`. |
+| LSIG_TEMPLATE      | Signature of an Algorand program that compiles with a known template |
```
Comment (deanstef) [inline ARCs/arc-0060.md:212]: After several IRL discussions, I favour removing `LSIG_TEMPLATE`.

File: ARCs/arc-0060.md [hunk 4]
```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
...
 
-#### Interface `StdSignMetadata`
+`SignDataError` uses the same error codes of `SignTxnsError` as well as the following codes:
 
-A `StdSignMetadata` object specifies the metadata for the `StdSigData`.
+| Status Code | Name | Description |
+| --- | --- | --- |
+| 4600 | Invalid scope | The scope of the signing action is not recognized by the wallet |
+| 4601 | Doesn't match schema | The data does not adhere to the scope schema |
+| 4602 | Failed parsing | Data cannot be parsed according to the schema |
```
Comment (deanstef) [inline ARCs/arc-0060.md:232]: What is the difference between `4600` and `4602`?

File: ARCs/arc-0060.md [hunk 5]
```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
...
 ```ts
-export interface SignDataError extends Error {
-  code: number;
-  data?: any;
-  failingSignData: (StdSigData | null);
+function parse(data: string): Uint8Array {
+    return Buffer.from(data, 'base64');
+}
+function validate(data: Uint8Array): boolean {
+    return data.length >= 2 && data[0] === 0x4D && data[1] === 0x58;
```
Comment (deanstef) [inline ARCs/arc-0060.md:288]: The length should be strictly greater than two IMO. I won't allow users to sign just the prefix `MX`.  WDYT?

File: ARCs/arc-0060.md [hunk 6]
```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
...
 
-| Status Code | Name | Description |
-| --- | --- | --- |
-| 4600 | Invalid scope | The scope of the signing action is not recognized by the wallet |
-| 4601 | Invalid schema | The schema does not comply with ARC-60 requirements |
-| 4602 | Invalid encoding | Data cannot be decoded with the given encoding |
+#### EXAMPLE Scope `DOG_OWNERSHIP`
 
-### Semantic Requirements
+This scope is intended purely as an example to showcase how a scope with a stricter schema can be signed. The `DOG_OWNERSHIP` scope signs a JSON object that states that a person owns a dog:
```
Comment (deanstef) [inline ARCs/arc-0060.md:299]: The problem I see with this approach is that we are not using JSON schema but JSON objects. 

Schemas have a particular structure (https://json-schema.org/understanding-json-schema/reference) that we should follow, e.g. to enforce data validation rules against the schemas.

File: ARCs/arc-0060.md [hunk 7]
```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
...
+    dogChipNumber: string;
+}
+```
 
-> Following [ARC-1](./arc-0001.md) terminology, in this ARC the term **Rejecting** means throwing an error with `err.code=4300`.
+We assume that a dog's chip number must be a 15-digit number and that the onwer name must be between 1 and 50 characters long. A `DogOwnership` is serialized according to canonical JSON ([RFC8785](https://www.rfc-editor.org/rfc/rfc8785)), which is intended to seralize JSON deterministically for use in cryptographic operations (hashing, signing, etc). We assume there exists a function `toCanonicalJson` that serializes an object to a byte array according to RFC8785.
 
-Upon calling `signData(signingData,, metadata)`:
+```ts
+function parse(data: string): DogOwnership {
```
Comment (deanstef) [inline ARCs/arc-0060.md:311]: I don't understand how the schema can "define" these functions though. IMO with a schema, and encoding, we have everything needed to decode and validate a serialized signing object.

What's the need for such functions?

File: ARCs/arc-0060.md [hunk 8]
```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
...
 - `data`:
-  - it **MUST** be a valid `StdDataStr` object (UTF-8 encoded), otherwise the wallet **MUST** reject.
-  - it **MUST** adhere to `metadata.schema`, otherwise the wallet **MUST** reject.
-  - the `data.ARC60Domain` field must be validated against forbidden values:
-    - the wallet **MUST** reject if `ARC60Domain="TX"`.
-    - the wallet **MUST** reject if `ARC60Domain="TG"`.
-  - `data.bytes` must be a valid byte string:
-    - it **MUST** be encoded as `metadata.encoding` if present, otherwise `base64`.
-    - the wallet **MUST** throw a `4602` error if the decoding fails with the given encoding value.
-    - the wallet **MUST** reject if `data.bytes` field is prepended with a forbidden domain separator `TX` or `TG`.
```
Comment (deanstef) [inline ARCs/arc-0060.md:348]: In this version of the ARC we never mention forbidden domain separators like `TX` or `TG`. We should do that unless we want people to start using ARC60 to sign valid transactions.

File: ARCs/arc-0060.md (hunk 1 above)
Comment (k13n) [reply ARCs/arc-0060.md:77]: I agree, I was using the terms somewhat interchangeably. I was thinking that an ARC defines a schema and within this ARC also the three functions are defined. E.g., ARC-31 proposes a scope (`AUTH`), its schema (the JSON schema that you defined), and how the data is to be parsed, validated, and serialized.

File: ARCs/arc-0060.md (hunk 3 above)
Comment (k13n) [reply ARCs/arc-0060.md:212]: agreed 👍

File: ARCs/arc-0060.md (hunk 4 above)
Comment (k13n) [reply ARCs/arc-0060.md:232]: Do you mean `4601` and `4602`? Good question if both are needed, probably not.

File: ARCs/arc-0060.md (hunk 5 above)
Comment (k13n) [reply ARCs/arc-0060.md:288]: makes sense! I was also wondering if we should change the way `MX_RANDOM` works to mirror what the Algorand SDK does (https://github.com/algorand/js-algorand-sdk/blob/develop/src/main.ts#L68).

In this PR we have:
- `validate`: check if `data` starts with `MX`
- `serialize`: returns `data` as is

If the do it more like the Algorand SDK does it, we'd have to do:
- `validate`: no-op
- `serialize`: `concat('MX', data)`

File: ARCs/arc-0060.md (hunk 6 above)
Comment (k13n) [reply ARCs/arc-0060.md:299]: Good point. I didn't want to limit the ARC to just use JSON objects as payload and JSON schema for schema definition. If an ARC (like ARC-31) defines `parse`, `validate`, `serialize` in terms of JSON objects and JSON schema, that's fine. For example, in ARC-31 this might be `parse=JSON.parse`, `validate=validate_with_json_schema` and `serialize=msgpack.encode`.

If there's an ARC that doesn't want to use JSON or JSON schema, they can define their own rules independent of any JSON notation. For example, this `CHALLENGE32` schema that we talked about doesn't really need a JSON schema. Or let's say a SIWE (EIP-4361) style text format also doesn't have a JSON structure / schema.

Do you think modeling everything in terms of JSON objects and JSON schemas would be beneficial?

File: ARCs/arc-0060.md (hunk 8 above)
Comment (k13n) [reply ARCs/arc-0060.md:348]: I think it's the schema's responsibility to make sure that the object being signed doesn't serialize to something that's a transaction, logicsig, etc.

For example, the `CHALLENGE32` scope that we talked about has 32-byte payloads and is too short to encode a valid transaction, hence it'd be okay if a payload starts with `TX` or `TG`. If we enforce that a payload cannot have `TX` or `TG` as prefix, we cannot support truly random payload (like a random nonce). It was my impression that this needs to be supported.

File: ARCs/arc-0060.md (hunk 7 above)
Comment (k13n) [reply ARCs/arc-0060.md:311]: > I don't understand how the schema can "define" these functions though.

Somebody puts forth an ARC that defines in plain English or with a reference implementation (preferably both) how the data is parsed, validated, and serialized.

> IMO with a schema, and encoding, we have everything needed to decode and validate a serialized signing object.

Don't you think there are validation rules that cannot be expressed in JSON schema? Like, does the `domain` in an ARC-31 field match the domain of the dApp that the wallet is connected to, etc.

> What's the need for such functions?

As a wallet provider I care about three things: (a) how do I make sense of the data that I get from a dApp, (b) how do I validate if the data is okay according to the specs, and (c) how do I turn the data into bytes such that I can feed them to a signing algorithm. Functions `parse`, `validate`, and `serialize` help me answer those questions.

But perhaps I'm over-complicating things and a JSON schema is sufficient. I don't know. I'd love to finally see a concrete example of structured data (and not just a simple byte array) that needs to be signed in Algorand. Maybe that would help settle this debate
